src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs

```diff
@@ -196,6 +207,77 @@ static ImmutableArray<AssetsFileLogMessage> ParseLogMessages(LockFile lockFile,
             }
 
             return builder.ToImmutable();
+
+            Dictionary<string, LogLevel> BuildLevelByLibrary()
+            {
+                Dictionary<string, HashSet<string>> ancestorsByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                // Build a map from child-to-ancestors
+                foreach (LockFileTargetLibrary lockFileLibrary in lockFileTarget.Libraries)
+                {
+                    string? parentId = lockFileLibrary.Name;
+
+                    if (parentId is null)
+                    {
+                        continue;
+                    }
+
+                    foreach (var dep in lockFileLibrary.Dependencies)
+                    {
+                        string childId = dep.Id;
+
+                        if (!ancestorsByLibrary.TryGetValue(childId, out HashSet<string>? ancestors))
+                        {
+                            ancestors = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                            ancestorsByLibrary.Add(childId, ancestors);
+                        }
+
+                        ancestors.Add(parentId);
+                    }
+                }
+
+                // Walk up the tree from each log message's library, propagating the highest level found.
+                Dictionary<string, LogLevel> levelByLibrary = new(StringComparer.OrdinalIgnoreCase);
+
+                foreach (AssetsFileLogMessage message in logMessages)
+                {
+                    Integrate(message.LibraryName, message.Level, visited: []);
+                }
+
+                return levelByLibrary;
+
+                void Integrate(string id, LogLevel level, HashSet<string> visited)
+                {
+                    if (!visited.Add(id))
+                    {
+                        // Avoid infinite recursion.
+                        return;
+                    }
+
+                    if (!levelByLibrary.TryGetValue(id, out LogLevel currentLevel))
+                    {
+                        // No level yet, so set it.
+                        levelByLibrary[id] = level;
+                    }
+                    else
+                    {
+                        // Higher level is more severe. If we already have a higher level, don't change it.
+                        // Also, if we already have a level, assume we have propagated it to all ancestors and can return.
```

Comment:
How can this assumption be made? Because the ordering of the messages in the assets file imply that ancestors come first?

Reply:
All nodes start with no level. Any time a level is assigned, it's pushed to all ancestors. So if, when walking, you find a node with a level, you know that it's already present on all ancestors.

Reply:
I modified the comment. Hopefully it's clearer now.